lib/lab_manager/provider/vsphere.rb

```diff
@@ -43,30 +55,133 @@ def create_vm_options
     # whould be nice to be able to validate before sendting a request
 
     def create_vm(opts = {})
-      opts = opts.reverse_merge(VSphereConfig.run_default_opts)
+      opts = opts.reverse_merge(VSphereConfig.create_vm_defaults || {})
       VSphere.connect.with do |vs|
-        vs.vm_clone(
-          'datacenter'    => opts[:datacenter],
-          'template_path' => template_uid,
-          'name'          => opts[:name],
-
-          'cluster'       => opts[:cluster],
-          'linked_clone'  => opts[:linked_clone],
-          'dest_folder'   => compute.dest_folder || default_dest_folder(opts)
+
+        dest_folder = opts[:dest_folder]
+        vm_name = opts[:name] || 'lm_' + SecureRandom.hex(8)
+        exception_cb = -> do
+          LabManager.logger.warn(
+            "Failed attempt to create virtual machine:  template_name: #{vm_name}"
+          )
+        end
+        Retryable.retryable(
+          tries: 3,
+          on: [RbVmomi::Fault, MachineCreationFailed],
+          exception_cb: exception_cb
+        ) do
+          @machine = vs.vm_clone(
+            'datacenter'    => opts[:datacenter],
+            'template_path' => opts[:template_path],
+            'name'          => vm_name,
+            'cluster'       => opts[:cluster] || 'kraken',
+            'linked_clone'  => opts[:linked_clone] || true,
+            'dest_folder'   => dest_folder,
+            'power_on'      => opts[:power_on] || false,
+            'wait'          => true
+          )
+
+          raise MachineCreationFailed, "MachineCreationFailed, retrying (#{vm_name})" unless @machine['vm_ref']
+          compute.provider_data = vm_data(@machine['new_vm'], vs: vs)
+        end
+
+        add_machine_to_drs_rule(
```

Comment:
call this method only if it is desirable by config, e.g. you may not need to set any drs_rule.